PWA updates

dodumosu opened the pull request:
fixes nditech/apollo-issues#114

dodumosu committed:
fix: add missing else clause

fixes nditech/apollo-issues#114

dodumosu committed:
fix: check for isUpdateAvailable after DOM load

this commit addresses nditech/apollo-issues#115

dodumosu committed:
fix: remove permissions check

dodumosu committed:
chore: convert arrow methods to functions

dodumosu commented:
also addresses nditech/apollo-issues#116 as far as submissions on iOS is concerned

dodumosu committed:
chore: change notification type and message

dodumosu committed:
chore: remove extraneous submit button

dodumosu committed:
feat: display range for integer fields

dodumosu committed:
feat: add background

dodumosu committed:
feat: only compare submission data

dodumosu committed:
feat: delay submission add

dodumosu committed:
feat: various changes

- replace zero delta with a small one (500ms) when checking
  for pending submissions
- remove "missing" icon (nditech/apollo-issues#117)
  use icon with description on colored badge
  for other completion statuses

dodumosu committed:
chore: minor refactor

dodumosu committed:
feat: add location access toggle

add a switch to toggle location access.
this will *NOT* stop the browser requesting access if it is on,
at least, at first.

see: nditech/apollo-issues#120

dodumosu committed:
feat: add label to connectivity indicator

see: nditech/apollo-issues#122

dodumosu committed:
fix: provide default if event is falsy

dodumosu committed:
chore: change button labels

- change 'Back' button label to 'Save and close'
- change 'Cancel' button label to 'Close without saving'

see: nditech/apollo-issues#125

dodumosu committed:
chore: reposition form type badge

see: nditech/apollo-issues#126

dodumosu committed:
fix: various fixes

- don't notify user about inability to get location if they have turned
  off location access
- method to persist submissions to IndexedDB no longer updates the
  `updated` field. callers need to update it instead.

see: nditech/apollo-issues#124

dodumosu committed:
feat: allow upload of logos

dodumosu committed:
feat: resize logo to 30x30

see: nditech/apollo-issues#123

dodumosu committed:
feat: highlight modified field labels

see: nditech/apollo-issues#127

dodumosu committed:
feat: add new endpoint to get forms and serials

this commit adds a new API endpoint to retrieve forms and form serials.

see: nditech/apollo-issues#121

dodumosu committed:
chore: change format for serials

dodumosu committed:
chore: clean up envelope

dodumosu committed:
feat: integrate new enpoint in PWA

dodumosu committed:
chore: change wording

dodumosu committed:
feat: change completion rendering for checklists

this commit introduces the following changes:
- location access toggle is now a checkbox, not a switch
- for checklists, each section is rendered as a badge on form list
- some minor fixes

see: nditech/apollo-issues#128

dodumosu committed:
feat: add autocomplete for form serials

dodumosu committed:
fix: check multiselect completion separately

dodumosu committed:
fix: completion for multiselect fields

dodumosu committed:
chore: update section header colour

dodumosu committed:
feat: only collapse completely filled out sections

dodumosu committed:
fix: add validation method

dodumosu committed:
chore: debounce keyboard input

dodumosu committed:
chore: minor cleanup

dodumosu committed:
fix: change key name

dodumosu committed:
chore: remove unnecessary code

dodumosu committed:
fix: suspend class bind

see: nditech/apollo-issues#130

dodumosu committed:
fix: reincorporate close functionality

see: nditech/apollo-issues#131

dodumosu committed:
feat: minor refactors

- reload database data when submission editor is closed
- remove diff source
- add minified fast-copy
- restore initial load settings for submission editor

dodumosu committed:
chore: remove "pending submissions" info box

see: nditech/apollo-issues#119

dodumosu committed:
feat: add errorFields key to error payload

dodumosu committed:
chore: minor refactor

dodumosu committed:
feat: various changes

- fixes an issue on checklist load
- reverts header colors when text in a field is cleared
- adds range validation for integer fields on submit

see: nditech/apollo-issues#135

dodumosu committed:
chore: extra value checks

dodumosu committed:
chore: minor cleanup

dodumosu committed:
fix: conditionally initialize autocomplete

dodumosu committed:
feat: change completion rendering

- remove label color change on edit
- section colors change based on status of posted data

see: nditech/apollo-issues#133

dodumosu committed:
chore: cleanup labels

dodumosu committed:
fix: restrict form query

dodumosu committed:
feat: submission status badge

see: nditech/apollo-issues#133

dodumosu committed:
chore: change error message

dodumosu committed:
feat: add section clear button

dodumosu committed:
chore: don't disable submit if browser is offline

dodumosu committed:
feat: remove form type

see: nditech/apollo-issues#141

dodumosu committed:
feat: revert to initial setup

dodumosu committed:
fix: switch to proxy object

this commit fixes the issue raised in nditect/apollo-issues#142

see: nditech/apollo-issues#142

dodumosu committed:
feat: exclude used form serials

this commit introduces the following changes:
- exclude used form serials from the autocomplete;
- the purge functionality now purges serials too.
  it is still unexposed by the UI

see: nditech/apollo-issues#138

dodumosu committed:
fix: don't prompt for serial on autosave

dodumosu committed:
chore: refactor and move `qa_status`

dodumosu committed:
feat: update returned data

this commit updates the response payload
on submission with the current data fields
and whether or not QA failed

see: nditech/apollo-issues#143

dodumosu committed:
feat: add QA failed icon

dodumosu committed:
chore: sort field options

see: nditech/apollo-issues#147

dodumosu committed:
chore: add metadata tags

dodumosu committed:
chore: clean up returned fields

dodumosu committed:
chore: update manifest

dodumosu committed:
chore: update image-blob-reduce

dodumosu committed:
chore: remove unminified and unused static files

dodumosu committed:
chore: cleanup

dodumosu committed:
chore: spruce up login form

dodumosu committed:
feat: only update updated flag in handlers

see: nditech/apollo-issues#146

dodumosu committed:
chore: add Sentry tracking to main PWA branch

dodumosu committed:
chore: formatting change

dodumosu committed:
feat: change layout

dodumosu committed:
feat: render two sets of statuses

dodumosu committed:
chore: minor refactor

dodumosu committed:
feat: add isComplete function

dodumosu committed:
feat: disable add button when all serials used

dodumosu committed:
chore: minor refactor

dodumosu committed:
chore: prune code

dodumosu committed:
feat: add form type badge, other changes

dodumosu committed:
feat: change online status during commit checks

dodumosu committed:
feat: upgrade database version

dodumosu committed:
fix: don't disable button if not survey form

dodumosu committed:
feat: mark checklist as done if done

dodumosu committed:
fix: restore auto-close for section headers

dodumosu committed:
feat: restore pending status

takinbo commented:
@dodumosu can we rename this PR appropriately?

takinbo commented:
Could you also rebase and resolve the conflicts?

takinbo commented on apollo/settings.py:
```diff
@@ -41,6 +41,8 @@
 
 SENTRY_DSN = config('SENTRY_DSN', default=None)
 SENTRY_USER_ATTRS = ['email']
+TRACK_PWA_ERRORS = config(
```
Can we avoid having a PWA specific error tracking setting? If `DEBUG` is true, then we should track PWA errors.

takinbo commented on apollo/pwa/static/js/app.js:
```diff
@@ -1,33 +1,29 @@
+  if ('serviceWorker' in navigator) {
+    navigator.serviceWorker.register('/pwa/serviceworker.js', {scope: '/pwa/'})
+      .then(registration => {
+        console.log('service worker registered successfully with scope:', registration.scope);
```
we want to remove these "debug" statements when the application gets to production

dodumosu committed:
chore: change background color

see: nditech/apollo-issues#149

dodumosu committed:
chore: remove unused argument

dodumosu committed:
fix: add missing import

dodumosu committed:
feat: implement queueing

dodumosu committed:
chore: update bootstrap to 5.0.0

dodumosu committed:
feat: add popper.js and tippy.js

dodumosu committed:
feat: add scripts to template & service worker

dodumosu committed:
feat: change editing layout

dodumosu committed:
chore: reduce card height

dodumosu committed:
fix: fix 'updated null' issue

dodumosu committed:
chore: minor refactor

dodumosu committed:
feat: change token lifetimes

- change access token lifetime to 72H
- remove setting for refresh token

dodumosu committed:
feat: change sync intervals

- form sync: every hour
- data sync: every minute
- version check: every two hours

dodumosu committed:
feat: remove close button for incidents

dodumosu committed:
chore: layout changes

dodumosu committed:
chore: more layout changes

dodumosu committed:
feat: change sync form to every quarter hour

dodumosu committed:
feat: add global status

dodumosu committed:
fix: workaround for Apple's inane mobile keyboard

see: nditech/apollo-issues#152

dodumosu committed:
chore: show login form on expired session

if session is expired, show login form

see: nditech/apollo-issues#150

dodumosu committed:
feat: add submission UUID in response

dodumosu committed:
feat: backend support for QA checks

dodumosu committed:
feat: add QA status checks

dodumosu committed:
fix: prompt cache update

dodumosu committed:
feat: consolidate database changes
